ci/src/main.rs

```diff
@@ -0,0 +1,213 @@
+            });
+        }
+
+        !(is_nightly || is_beta)
+    };
+
+    // Sanity check, do this up front so we fail early.
+    run_cmd!(cargo --version)?;
+    run_cmd!(rustc --version)?;
+
+    // Help troubleshoot CI runs, display the status of all features if `--debug` is set.
+    display_features(&args);
+
+    // "std" and "no-std" tests, examples, and feature matrix.
+    //
+    // Since rust-bitcoin requires either "std" or "no-std" to be enabled we group the test
+    // accordingly. If `--feature-matrix` is set we test each of `FEATURES` coupled with std/no-std.
+    if std {
+        debug("Build and test the default features ...");
+        run_cmd!(cargo build)?;
+        run_cmd!(cargo test)?;
+
+        if args.feature_matrix {
+            debug("Build and test with no features other than std ...");
+            run_cmd!(cargo build --no-default-features --features="std")?;
+            run_cmd!(cargo test --no-default-features --features="std")?;
+
+            debug("Build and test individual features along with \"std\":");
+            for feature in FEATURES.split(" ") {
+                let features = format!("std {}", feature);
+                if args.debug {
+                    eprintln!("\t feature \"{}\"", feature);
+                }
+                run_cmd!(cargo test --no-default-features --features="$features")?;
+            }
+
+            debug("Build std + no_std, to make sure they are not incompatible ...");
+            run_cmd!(cargo build --no-default-features --features="std no-std")?;
+        }
+
+        debug("Run the examples with the \"std\" feature enabled ...\n");
+        run_cmd!(cargo run --example bip32 7934c09359b234e076b9fa5a1abfd38e3dc2a9939745b7cc3c22a48d831d14bd)?;
+
+    } else {
+        debug("Build no_std, to make sure that cfg(test) doesn't hide any issues ...");
+        run_cmd!(cargo build --no-default-features --features="no-std")?;
+
+        debug("Test with no features other than \"no-std\" ...");
+        run_cmd!(cargo test --no-default-features --features="no-std")?;
+
+        debug("Build all the features excluding \"std\" ...");
+        let features = format!("no-std {}", FEATURES);
+        run_cmd!(cargo test --no-default-features --features="$features")?;
+
+        if args.feature_matrix {
+            debug("Build and test individual features along with \"no-std\" ...");
+            for feature in FEATURES.split(" ") {
+                let features = format!("no-std {}", feature);
+                run_cmd!(cargo test --no-default-features --features="$features")?;
+            }
+        }
+
+        debug("Run the examples without \"std\" and with \"no-std\"  ...");
+        run_cmd!(cargo run --no-default-features --features no-std --example bip32 7934c09359b234e076b9fa5a1abfd38e3dc2a9939745b7cc3c22a48d831d14bd)?;
+    }
+
+    if args.docs {
+        if is_stable_toolchain {
+            eprintln!("docs build cannot be doe with a stable toolchain");
```

Comment:
nit: `s/doe/done`

Reply:
Spell checkers should use context already like word-processors do.